src/librustc_mir/dataflow/impls/borrows.rs

```diff
@@ -273,6 +273,246 @@ impl<'a, 'gcx, 'tcx> Borrows<'a, 'gcx, 'tcx> {
+
+            /// Returns true if the given location contains an NLL-activating use of the given place
+            fn location_contains_use(&self, location: Location, place: &Place) -> bool {
+                let mut use_checker = ContainsUseOfPlace::new(place);
+                let block = &self.mir.basic_blocks().get(location.block).unwrap_or_else(|| {
+                    panic!("could not find block at location {:?}", location);
+                });
+                if location.statement_index != block.statements.len() {
+                    // This is a statement
+                    let stmt = block.statements.get(location.statement_index).unwrap_or_else(|| {
+                        panic!("could not find statement at location {:?}");
+                    });
+                    use_checker.visit_statement(location.block, stmt, location);
+                } else {
+                    // This is a terminator
+                    match block.terminator {
+                        Some(ref term) => {
+                            use_checker.visit_terminator(location.block, term, location);
+                        }
+                        None => {
+                            // There is no way for Place to be used by the terminator if there is no
+                            // terminator
+                        }
+                    }
+                }
+
+                use_checker.use_found
+            }
+
+            /// Determines if the provided region is terminated after the provided location.
+            /// EndRegion statements terminate their enclosed region::Scope.
+            /// We also consult with the NLL region inference engine, should one be available
+            fn region_terminated_after(&self, region: Region<'tcx>, location: Location) -> bool {
+                let block_data = &self.mir[location.block];
+                if location.statement_index != block_data.statements.len() {
+                    let stmt = &block_data.statements[location.statement_index];
+                    if let mir::StatementKind::EndRegion(region_scope) = stmt.kind {
+                        if &ReScope(region_scope) == region {
+                            // We encountered an EndRegion statement that terminates the provided region
+                            return true;
+                        }
+                    }
+                }
+                if let Some(ref regioncx) = self.nonlexical_regioncx {
+                    if !regioncx.region_contains_point(region, location) {
+                        // NLL says the region has ended already
+                        return true;
+                    }
+                }
+
+                false
+            }
+
+            /// Computes the activation location of a borrow.
+            /// The general idea is to start at the beginning of the region and perform a DFS
+            /// until we exit the region, either via an explicit EndRegion or because NLL tells
```

Comment:
oh wow, you got this working *without* NLL, and just lexical regions via `EndRegion` ? How cool!

Reply:
(or maybe you don't have `EndRegion` support completely baked yet? I saw evidence of some XXX-marked code that came and went...)

Reply:
It does work without NLL; since we're walking the MIR statement by statement and only considering one region at a time we can just look for the explicit `EndRegion` instead of asking the NLL region inference engine where the region ends. The code that came and went was me playing with the idea of handling all kills for borrows going out of scope in `kill_borrows_out_of_scope_at_location` but it ended up being cleaner to leave the existing implementation (where `EndRegion` gets handled in the big `match` in `statement_effect`) IMO.